### PR TITLE
Symplectic Grassmann

### PR DESCRIPTION
**Pull request by kellertuer**
this PR is a start to implement the (real) Symplectic Grassmann, which one obtains from Symplectic Stiefel the same way one obtains Grassmann from Stiefel.
It will probably also revise a few symplectic Stiefel functions.

* [x] Introduce `Hamiltonian`, they allow for `H^+` to be nicely defined for the symplectic inverse 
* [x] We should introduce `HamiltonianMatrices`? Maybe not 100% necessary but nice for completeness
* [x] Define `SymplecticGrassmann`
* [x] implement checks
* [x] `inner`  (after (4.10) / Lemma 4.8, BZ21)
* [x] the `test/utils.jl` functions complain since Julia 1.10 to be multiply included. We should at least move those, if not also the constants to the Manufolds.jl/Test.jl extension?
* [x] unify notation concerning J_2n and Q_2n for the symplectic element (and refer to `SymplecticMatrix`).
* [x] unify/simplify notation to refer more often to `symplectic_inverse` when writing `p^+` instead of repeating the formula.
* [x] implement the geodesic
* [x] implement the Cayley retraction and its inverse 
* [x] refactor `rand_hamiltonian` to be `rand(Hamiltonian(2n))`
* [x] implement the gradient conversion (after Lemma 4.8, BZ21)
* [x] (maybe) check for `Symplectic` and `SymplecticStiefel`, so they get `rand!` methods as well. They currently just have `rand` methods
* [x] read docs carefully again
* [x] check test coverage

Both symplectic Stiefel and Grassmann stem from this paper https://arxiv.org/abs/2108.12447

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -37,27 +43,27 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 * Improved distribution of random vector generation for rotation matrices and complex circle.
 
-## [0.9.7] - 2023-11-14
+## [0.9.7] * 2023-11-14
```
Comment (mateuszbaran) [inline NEWS.md:46]: Why did you change `-` to `*`?

Comment (kellertuer) [reply NEWS.md:46]: oh that was by accident. I changed itemise points to a unified format, I think/hope. Will check that.

File: docs/make.jl [hunk 2]
```diff
@@ -113,6 +113,7 @@ makedocs(;
                 "Generalized Stiefel" => "manifolds/generalizedstiefel.md",
                 "Generalized Grassmann" => "manifolds/generalizedgrassmann.md",
                 "Grassmann" => "manifolds/grassmann.md",
+                "Hamiltonian" => "manifolds/hamiltonian.md",
```
Comment (mateuszbaran) [inline docs/make.jl:116]: Is it a separate manifold?

Comment (mateuszbaran) [reply docs/make.jl:116]: I don't see a separate `<:AbstractManifold` there.

Comment (kellertuer) [reply docs/make.jl:116]: Not yet. See list above. It just makes sense for consistency to introduce it, it will be super similar to symmetric (a Euclidean sub manifold with the Euclidean metric).

Comment (kellertuer) [reply docs/make.jl:116]: Unless you are against that, but my two comparison manifolds are symmetric and skew symmetric – not super spectacular manifolds, but sometimes nice to have. The same holds here.

Comment (mateuszbaran) [reply docs/make.jl:116]: This is completely fine.

Comment (kellertuer) [reply docs/make.jl:116]: Ok, I will add that then, again, not super spectacular, but maybe useful for someone.

**Comment by kellertuer** [comment]
Hmpf, just triggered a few links from a wrong branch that I created yesterday for a test – not my day today.

**Comment by kellertuer** [comment]
With the last few methods now I would consider this feature-complete (just not yet test&docs complete), so I moved it from draft to ready (but did not yet set the label).

File: NEWS.md [hunk 3]
```diff
@@ -5,6 +5,18 @@ All notable changes to this project will be documented in this file.
...
+
+### Added
+
+* added the real symplectic Grassmann manifold `SymplecticGrassmann`
+* Introduce the manifold of `HamiltonianMatrices` and a wrapper for `Hamiltonian` matrices
+* introduce `rand(:HamiltonianMatrices)`
+* extend `rand` to also `rand!` for `HamiltonianMatrices`, `SymplecticMatrices` and `SymplecticStiefel`
+* implement `riemannian_gradient` conversion for `SymplecticMatrices` and `SymplecticGrassmann`
+* Rename (deprecate) `Symplectic` to `SimplecticMatrices` in order to have a `Symplectic` wrapper for such matrices as well
+* Rename (deprecate) `SymplecticMatrix` to `SymplecticElement` to clarify that it is the special matrix ``J_{2n}`` and not an arbitrary symplectic matrix.
```
Comment (mateuszbaran) [inline NEWS.md:18]: These should be in a "Deprecated" section.

File: docs/src/features/atlases.md [hunk 4]
```diff
@@ -11,9 +11,9 @@ For an atlas ``\mathcal A`` we further require that
 ```
 
 We say that ``φ_i`` is a chart about ``p``, if ``p\in U_i``.
-An atlas provides a connection between a manifold and the Euclidean space ``\mathbb{R}^n``, since
+An atlas provides a connection between a manifold and the Euclidean space ``ℝ^n``, since
 locally, a chart about ``p`` can be used to identify its neighborhood (as long as you stay in ``U_i``) with a subset of a Euclidean space.
-Most manifolds we consider are smooth, i.e. any change of charts ``φ_i \circ φ_j^{-1}: \mathbb{R}^n\to\mathbb{R}^n``, where ``i,j\in I``, is a smooth function. These changes of charts are also called transition maps.
+Most manifolds we consider are smooth, i.e. any change of charts ``φ_i \circ φ_j^{-1}: ℝ^n\toℝ^n``, where ``i,j\in I``, is a smooth function. These changes of charts are also called transition maps.
```
Comment (mateuszbaran) [inline docs/src/features/atlases.md:16]: It might not render correctly without a space after `\to`.

File: docs/src/manifolds/fiber_bundle.md [hunk 5]
```diff
@@ -4,7 +4,7 @@ Fiber bundle $E$ is a manifold that is built on top of another manifold $\mathca
 It is characterized by a continuous function $Π : E → \mathcal M$. For each point $p ∈ \mathcal M$ the preimage of $p$ by $Π$, $Π^{-1}(\{p\})$ is called a fiber $F$.
 Bundle projection can be performed using function [`bundle_projection`](@ref).
 
-`Manifolds.jl` primarily deals with the case of trivial bundles, where $E$ can be identified with a product $M \times F$.
+`Manifolds.jl` primarily deals with the case of trivial bundles, where $E$ can be identified with a product $M ×F$.
```
Comment (mateuszbaran) [inline docs/src/manifolds/fiber_bundle.md:7]:
```suggestion
`Manifolds.jl` primarily deals with the case of trivial bundles, where $E$ can be topologically identified with a product $M×F$.
```
A small omission I've noticed while looking at it.

File: src/manifolds/GeneralizedGrassmann.jl [hunk 6]
```diff
@@ -127,7 +127,7 @@ end
     distance(M::GeneralizedGrassmann, p, q)
 
 Compute the Riemannian distance on [`GeneralizedGrassmann`](@ref)
-manifold `M`$= \mathrm{Gr}(n,k,B)$.
+manifold `M```= \mathrm{Gr}(n,k,B)``.
```
Comment (mateuszbaran) [inline src/manifolds/GeneralizedGrassmann.jl:130]: Does this work without some spaces?

File: src/manifolds/QuotientManifold.jl [hunk 7]
```diff
@@ -172,11 +175,14 @@ function Base.show(io::IO, M::QuotientManifold)
     return print(io, "QuotientManifold($(M.manifold), $(M.total_space))")
 end
 
-"""
+@doc raw"""
     vertical_component(N::AbstractManifold, p, X)
+    vertical_compontent(QuotientManifold{M,N}, p, X)
```
Comment (mateuszbaran) [inline src/manifolds/QuotientManifold.jl:180]:
```suggestion
    vertical_component(QuotientManifold{M,N}, p, X)
```

File: test/ambiguities.jl [hunk 8]
```diff
@@ -1,3 +1,14 @@
+"""
+    has_type_in_signature(sig, T)
+    Test whether the signature `sig` has an argument of type `T` as one of its paramaters
```
Comment (mateuszbaran) [inline test/ambiguities.jl:3]:
```suggestion
    has_type_in_signature(sig, T::Type)

Test whether the signature `sig` has an argument of type `T` as one of its parameters.
```

File: test/groups/group_utils.jl [hunk 9]
```diff
@@ -1,31 +1,35 @@
-struct NotImplementedOperation <: AbstractGroupOperation end
+s = @isdefined _group_utils_included
+if !s
+    _group_utils_included = true
```
Comment (mateuszbaran) [inline test/groups/group_utils.jl:3]: This reminds me of the `#ifndef / #define / #endif` idiom in C and C++ :smile: . We need `#pragma once` in Julia.

**Review by mateuszbaran — COMMENTED** [review]
I haven't read the actual new stuff but here are some small technical comments.

File: NEWS.md (hunk 3 above)
Comment (kellertuer) [reply NEWS.md:18]: Well-spotted, will check with the official guideline we follow.

File: docs/src/features/atlases.md (hunk 4 above)
Comment (kellertuer) [reply docs/src/features/atlases.md:16]: That is probably correct, yes. But the `\to` should/could also become its actual error symbol then, Will  search/replace that as well then.

File: src/manifolds/GeneralizedGrassmann.jl (hunk 6 above)
Comment (kellertuer) [reply src/manifolds/GeneralizedGrassmann.jl:130]: oh, I am not sure, if it would, that would of course be great, but maybe it does not.

File: test/groups/group_utils.jl (hunk 9 above)
Comment (kellertuer) [reply test/groups/group_utils.jl:3]: this is for now also a bit of a quick hack. I learned that Julia 1.9 also did not like multiple includes, but was silent about it, better would be a small test package or such, but for now this reduces the output during testing _tremendously_

File: NEWS.md (hunk 3 above)
Comment (kellertuer) [reply NEWS.md]: You are right, it also reads better that way. Fixed it.

**Comment by kellertuer** [comment]
Uff, next time I have to start tests along the way as well, spent much too much time on these today. Now just the projector representation is missing, but since we have no functions for these besides the checks, the only thing to come of with is a point ant a tangent vector.

**Comment by kellertuer** [comment]
Hm, I just wanted to be nice and define Matrix Multiplications for Hamiltonians, but that introduces 100+ ambiguities :( (I will also lift the boundary for the ambiguity check to include 1.10). What shall we do there best? Not define it for Hamiltonian * abstract Matrix? I will check what Symmetric does.

**Comment by mateuszbaran** [comment]
Yes, I think you should only define +-* for Hamiltonian with itself. Alternatively, you can define `getindex` and rely on the generic definitions.

**Comment by kellertuer** [comment]
Ah yes in between I got a few `getindex` errors, but they looked super involved, that it was unclear to me, which variants need to be implemented. For now this should be enough for the start at least.

**Comment by kellertuer** [comment]
In principle, this is just missing my understanding how to get from a StiefelTVector on SpGr to a ProjectorTVector (to be precise to the specific representation we use).

**Comment by kellertuer** [comment]
Today seems to be the day of infrastructure failure.

File: docs/src/manifolds/symplectic.md [hunk 10]
```diff
@@ -1,35 +1,35 @@
...
+``\operatorname{Sp}(2n, 𝔽)`` is a subgroup of the special linear group, ``\operatorname{SL}(2n, 𝔽)``. When the underlying
+field is either ``ℝ`` or ``ℂ`` the symplectic group with a manifold structure constitutes a Lie group, with the Lie
 Algebra
 ````math
-    \mathfrak{sp}(2n,F) = \{H \in \mathbb{F}^{2n \times 2n} \;|\; Q H + H^{T} Q = 0\}.
+    \mathfrak{sp}(2n,F) = \{H \in 𝔽^{2n×2n} \;|\; Q H + H^{T} Q = 0\}.
 ````
 This set is also known as the [Hamiltonian matrices](https://en.wikipedia.org/wiki/Hamiltonian_matrix), which have the
-property that $(QH)^T = QH$ and are commonly used in physics.
+property that $(QH)^T = QH`` and are commonly used in physics.
```
Comment (mateuszbaran) [inline docs/src/manifolds/symplectic.md:32]:
```suggestion
property that ``(QH)^T = QH`` and are commonly used in physics.
```

File: src/manifolds/QuotientManifold.jl [hunk 11]
```diff
@@ -150,18 +150,21 @@ end
...
-Compute the [`horizontal_lift`](@ref) of `X` from ``T_p\mathcal M``, ``p=π(q)``.
+Compute the horizontal lift of `X` from ``T_p\mathcal M``, ``p=π(q)``.
 to ``T_q\mathcal N` in place of `Y`.
 """
 horizontal_lift!(N::AbstractManifold, Y, q, X)
 
-"""
+@doc raw"""
     horizontal_component(N::AbstractManifold, p, X)
+    horizontal_compontent(QuotientManifold{M,N}, p, X)
```
Comment (mateuszbaran) [inline src/manifolds/QuotientManifold.jl:162]:
```suggestion
    horizontal_compontent(QuotientManifold{𝔽,M,N}, p, X)
```

File: src/manifolds/Hamiltonian.jl [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+        @assert size(A, 1) == 2 * n "The first dimension of A ($(size(A,1))) is not even"
+        @assert size(A, 2) == 2 * n "The matrix A is of size ($(size(A))), which is not square."
+        return new{T,S}(A)
+    end
+end
+# Avoid double wrapping / unwrap if that happened
+Hamiltonian(A::Hamiltonian) = Hamiltonian(A.value)
+# Conversion
+function Matrix(A::Hamiltonian)
+    return copy(A.value)
```
Comment (mateuszbaran) [inline src/manifolds/Hamiltonian.jl:28]:
```suggestion
    return Matrix(A.value)
```

File: src/manifolds/Hamiltonian.jl [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+    p2 = @view(pX[1:n, (n + 1):(2n)])
+    p3 = @view(pX[(n + 1):(2n), 1:n])
+    p4 = @view(pX[(n + 1):(2n), (n + 1):(2n)])
+    randn!(rng, p1)
+    p4 .= -p1'
+    randn!(rng, p2)
+    randn!(rng, p3)
+    p2 .= (1 / 2) .* (p2 .+ p2')
+    p3 .= (1 / 2) .* (p2 .+ p2')
+    pX .= (σ / norm(pX, 2)) .* pX
```
Comment (mateuszbaran) [inline src/manifolds/Hamiltonian.jl:208]: Isn't sigma supposed to dictate standard deviation of the distribution, not the exact norm of the matrix? I'd just multiply ``pX`` by sigma.

Comment (kellertuer) [reply src/manifolds/Hamiltonian.jl:208]: I think I really just rewrote the existing formula.

Comment (mateuszbaran) [reply src/manifolds/Hamiltonian.jl:208]: Then the existing formula was also incorrect.

Comment (kellertuer) [reply src/manifolds/Hamiltonian.jl:208]: Maybe, I can check and fix that.

File: docs/src/manifolds/symplectic.md (hunk 10 above)
Comment (kellertuer) [reply docs/src/manifolds/symplectic.md:32]: Well-spotted, maybe my regular expressions where a bit imprecise on that one.

**Comment by kellertuer** [comment]
This now covers all I wanted to be included. I already bumped the version, but I would also like to include the Multinomial PR (#702) in this version I think, since that one is basically finished as well.

File: src/manifolds/Hamiltonian.jl (hunk 13 above)
Comment (kellertuer) [reply src/manifolds/Hamiltonian.jl]: Fixed it as you proposed.

**Comment by mateuszbaran** [comment]
It looks like tolerance in one test for SymplecticStiefel is too low.

File: src/manifolds/Hamiltonian.jl [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    return print(io, "HamiltonianMatrices($(2n), $(F))")
+end
+function Base.show(io::IO, M::HamiltonianMatrices{Tuple{Int},F}) where {F}
+    n = get_parameter(M.size)[1]
+    return print(io, "HamiltonianMatrices($(2n), $(F); parameter=:field)")
+end
+
+@doc raw"""
+    p = rand(M::HamiltonianMatrices; σ::Real=1.0, vector_at=nothing)
+    rand(M::HamiltonianMatrices; σ::Real=1.0, vector_at=nothing)
```
Comment (mateuszbaran) [inline src/manifolds/Hamiltonian.jl:175]: Why is this doubled?

File: src/manifolds/Hamiltonian.jl [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+
+Generate a Hamiltonian matrix. Since these are a submanifold of ``ℝ^{2n×2n}``,
+the same method applies for points and tangent vectors.
+
+The generation is based on generating one normally-distributed
+``n×n`` matrix ``A`` and two symmetric ``n×n`` matrices ``B,C`` to generate
```
Comment (mateuszbaran) [inline src/manifolds/Hamiltonian.jl:181]:
```suggestion
The construction is based on generating one normally-distributed
``n×n`` matrix ``A`` and two symmetric ``n×n`` matrices ``B, C`` which are then stacked:
```
The word "generate" was overused here.

File: src/manifolds/SymplecticGrassmann.jl [hunk 16]
```diff
@@ -0,0 +1,120 @@
+@doc raw"""
+    SymplecticGrassmann{T,𝔽} <: AbstractEmbeddedManifold{𝔽, DefaultIsometricEmbeddingType}
+
+The symplectic Grassmann manifold consists of all symplectic subspaces of
+``ℝ^{2n}`` of dimension ``2k``, ``n ≥ k``.
+
+This manifold can be represented as corresponding representers on the [`SymplecticStiefel`](@ref)
```
Comment (mateuszbaran) [inline src/manifolds/SymplecticGrassmann.jl:7]:
```suggestion
Points on this manifold can be represented as corresponding representers on the [`SymplecticStiefel`](@ref)
```

File: src/manifolds/SymplecticGrassmann.jl [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+See also [`ProjectorPoint`](@ref) and [`StiefelPoint`](@ref) for these two representations,
+where arrays are interpreted as those on the Stiefel manifold.
+
+With respect to the quotient structure, the canonical projection ``π = π_{\mathrm{SpSt},\mathrm{SpGr}}`` is given by
+
+```math
+π: \mathrm{SpSt}(2n2k) → \mathrm{SpGr}(2n,2k), p ↦ π(p) = pp^+.
+```
+
+The tangent space is either the tangent space from the symplecti Stiefel manifold, where
```
Comment (mateuszbaran) [inline src/manifolds/SymplecticGrassmann.jl:29]:
```suggestion
The tangent space is either the tangent space from the symplectic Stiefel manifold, where
```

File: src/manifolds/SymplecticGrassmannStiefel.jl [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+function get_embedding(M::SymplecticGrassmann{Tuple{Int,Int},𝔽}) where {𝔽}
+    n, k = get_parameter(M.size)
+    return SymplecticStiefel(2n, 2k, 𝔽; parameter=:field)
+end
+
+@doc raw"""
+    inner(::SymplecticGrassmann, p, X, Y)
+
+Compute the Riemannian inner product ``g^{\mathrm{SpGr}}_p(X,Y)``, where ``p``
+is a point on the [`SymplecticStiefel`](@ref) manifold and ``X,Y \in \mathrm{Hor}_p^π\operatorname{SpSt}(2n,2k)``
```
Comment (mateuszbaran) [inline src/manifolds/SymplecticGrassmannStiefel.jl:69]: I think this should be Grassmann, not Stiefel?

**Comment by kellertuer** [comment]
I am already running this locally. I am a bit surprised by that since I did not change that.

File: src/manifolds/SymplecticGrassmannStiefel.jl [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+
+```math
+  X = J_{2n}^THJ_{2k}p^{\mathrm{T}}p - J_{2n}^TpJ_{2k}H^{\mathrm{T}}p,
+```
+
+where ``J_{2n}`` denotes the [`SymplecticElement`](@ref), and
+``H = (I_{2n} - pp^+)J_{2n}^{\mathrm{T}}YJ``.
+"""
+function riemannian_gradient(M::SymplecticGrassmann, p, Y; kwargs...)
+    n, k = get_parameter(M.size)
```
Comment (mateuszbaran) [inline src/manifolds/SymplecticGrassmannStiefel.jl:142]: Why do we need to implement both mutating and non-mutating variant?

**Review by mateuszbaran — COMMENTED** [review]
A few comments.

File: src/manifolds/Hamiltonian.jl (hunk 14 above)
Comment (kellertuer) [reply src/manifolds/Hamiltonian.jl:175]: Oh the second one is meant to be in-place `rand!(M[...], p)` and for consistency we should use `pX`

File: src/manifolds/SymplecticGrassmannStiefel.jl (hunk 18 above)
Comment (kellertuer) [reply src/manifolds/SymplecticGrassmannStiefel.jl:69]: the `SymplecticStiefel`? Complicated ;) Similar to Stiefel, when you represent SpGr in SpSt-points (think of similar to p on Grassman represented by one of its ONBs) then this correct in that form, but one could add, that p is a SpGr _represented_ in a SpSt. (other variant: represent p by its projector)

File: src/manifolds/Hamiltonian.jl [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+
+"""
+    is_flat(::HamiltonianMatrices)
+
+Return true. [`HamiltonianMatrices`](@ref) is a flat manifold.
+"""
+is_flat(M::HamiltonianMatrices) = true
+
+@doc raw"""
+    is_hamiltonian(A; kwargs...)
```
Comment (mateuszbaran) [inline src/manifolds/Hamiltonian.jl:145]:
```suggestion
    is_hamiltonian(A::AbstractMatrix; kwargs...)
```

File: src/manifolds/SymplecticGrassmannStiefel.jl (hunk 18 above)
Comment (mateuszbaran) [reply src/manifolds/SymplecticGrassmannStiefel.jl:69]: I see, I think adding the comment about representing in SpSt would make it more clear.

File: src/manifolds/Hamiltonian.jl [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+@doc raw"""
+    HamiltonianMatrices{T,𝔽} <: AbstractDecoratorManifold{𝔽}
+
+The [`AbstractManifold`](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/types.html#ManifoldsBase.AbstractManifold)
+consisting of (real-valued) hamiltonian matrices of size ``n×n``, i.e. the set
+
+````math
+\mathfrak{sp}(2n,𝔽) = \bigl\{p  ∈ 𝔽^{2n×2n}\ \big|\ p^+ = p \bigr\},
+````
+where ``⋅^{+}`` denotes the [`symplectic_inverse`](@ref),. and ``𝔽 ∈ \{ ℝ, ℂ\}``.
```
Comment (mateuszbaran) [inline src/manifolds/Hamiltonian.jl:49]:
```suggestion
where ``⋅^{+}`` denotes the [`symplectic_inverse`](@ref), and ``𝔽 ∈ \{ ℝ, ℂ\}``.
```

File: src/manifolds/Hamiltonian.jl [hunk 22]
```diff
@@ -0,0 +1,211 @@
+#
+# This file requires Symplectic to be defined, since it need the symplectic inverse A^+
+# This type is used in the Symplectic Grassmann
+@doc raw"""
+    Hamiltonian{T,S<:AbstractMatrix{<:T}} <: AbstractMatrix{T}
+
+A type to store an hamiltonien matrix, i.e. A square matrix matrix for which ``A^+ = -A`` where
```
Comment (mateuszbaran) [inline src/manifolds/Hamiltonian.jl:7]:
```suggestion
A type to store a Hamiltonian matrix, that is a square matrix for which ``A^+ = -A`` where
```

File: src/manifolds/Hamiltonian.jl [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+consisting of (real-valued) hamiltonian matrices of size ``n×n``, i.e. the set
+
+````math
+\mathfrak{sp}(2n,𝔽) = \bigl\{p  ∈ 𝔽^{2n×2n}\ \big|\ p^+ = p \bigr\},
+````
+where ``⋅^{+}`` denotes the [`symplectic_inverse`](@ref),. and ``𝔽 ∈ \{ ℝ, ℂ\}``.
+
+Though it is slightly redundant, usually the matrices are stored as ``2n×2n`` arrays.
+
+The symbol refers to the main usage within `Manifolds.jl` that is the
```
Comment (mateuszbaran) [inline src/manifolds/Hamiltonian.jl:53]: This is unclear. What symbol do you mean?

File: src/manifolds/Hamiltonian.jl [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+function Base.show(io::IO, M::HamiltonianMatrices{Tuple{Int},F}) where {F}
+    n = get_parameter(M.size)[1]
+    return print(io, "HamiltonianMatrices($(2n), $(F); parameter=:field)")
+end
+
+@doc raw"""
+    pX = rand(M::HamiltonianMatrices; σ::Real=1.0, vector_at=nothing)
+    rand!(M::HamiltonianMatrices, pX; σ::Real=1.0, vector_at=nothing)
+
+Generate a Hamiltonian matrix. Since these are a submanifold of ``ℝ^{2n×2n}``,
```
Comment (mateuszbaran) [inline src/manifolds/Hamiltonian.jl:177]:
```suggestion
Generate a random Hamiltonian matrix. Since these are a submanifold of ``ℝ^{2n×2n}``,
```

File: src/manifolds/SymplecticGrassmann.jl [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+tangent vectors are representers of their corresponding congruence classes, or for the
+representation as projectors, using a [`ProjectorTVector`](@ref) as
+
+```math
+  T_p\operatorname{SpGr}(2n, 2k, ℝ) =
+  \bigl\{ [X,p] \ \mid\ X ∈ \mathfrak{sp}(2n,ℝ), Xp+pX = X \bigr\},
+```
+
+where ``[X,p] = Xp-pX`` denotes the matrix commutator and
+``\mathfrak{sp}(2n,ℝ)`` is the Lie algebra of the symplectic group consisting of [`HamiltonianMatrices`](@ref).
```
Comment (mateuszbaran) [inline src/manifolds/SymplecticGrassmann.jl:39]: I don't think we actually have the symplectic group as a group in Manifolds.jl? Maybe we should have an issue open about it.

File: src/manifolds/SymplecticGrassmannStiefel.jl [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+    retract_cayley!(SymplecticStiefel(2 * n, 2 * k), q, p, X, t)
+    return q
+end
+
+@doc raw"""
+    riemannian_gradient(M::SymplecticGrassmann, p, Y)
+
+Given a gradient ``Y = \operatorname{grad} \tilde f(p)`` in the embedding ``ℝ^{2n×2k}`` or at
+least around the [`SymplecticGrassmann`](@ref) `M` where `p` (the embedding of) a point on `M`,
+and the restriction ``\tilde f`` to the [`SymplecticStiefel`] be invariant for the equivalence classes.
```
Comment (mateuszbaran) [inline src/manifolds/SymplecticGrassmannStiefel.jl:130]:
```suggestion
and the restriction ``\tilde f`` to the [`SymplecticStiefel`](@ref) be invariant for the equivalence classes.
```

File: src/manifolds/SymplecticGrassmannStiefel.jl [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+In other words ``f(p) = f(qp)`` for ``q \in \mathrm{Sp}(2k, ℝ)``.
```
Comment (mateuszbaran) [inline src/manifolds/SymplecticGrassmannStiefel.jl:131]: Is Sp(2k, R) defined somewhere?

File: src/manifolds/SymplecticGrassmannStiefel.jl [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+    n, k = get_parameter(M.size)
+    J = SymplecticElement(p, X, Y) # in BZ21 also J
+    # Procompute lu(p'p) since we solve a^{-1}* 3 times
+    a = lu(p' * p) # note that p'p is symmetric, thus so is its inverse c=a^{-1}
+    # we split the original trace into two one with I -> (X'Yc)
+    # 1) we permute X' and Y c to c^{\mathrm{T}}Y^{\mathrm{T}}X = a\(Y'X) (avoids a large interims matrix)
+    # 2) the second we compute as c (X'p)(p^+Y) since both brackets are the smaller matrices
+    return tr(a \ (Y' * X)) - tr(
+        a \ ((X' * p) * symplectic_inverse_times(SymplecticStiefel(2 * n, 2 * k), p, Y)),
+    )
```
Comment (mateuszbaran) [inline src/manifolds/SymplecticGrassmannStiefel.jl:89]: Could you verify if this can be rewritten as:
```julia
return dot(a \ Y'), X) - dot(
        a \ (X' * p),  symplectic_inverse_times(SymplecticStiefel(2 * n, 2 * k), p, Y),
    )
```
(I've used tr(X*Y) = dot(X, Y) and associativity).

File: src/manifolds/SymplecticGrassmannStiefel.jl (hunk 27 above)
Comment (kellertuer) [reply src/manifolds/SymplecticGrassmannStiefel.jl:131]: We could add that these are the Symplectic matrices.

File: src/manifolds/SymplecticGrassmannStiefel.jl (hunk 19 above)
Comment (kellertuer) [reply src/manifolds/SymplecticGrassmannStiefel.jl:142]: Then we could remove this, but I will only do that tomorrow (left my office already)

File: src/manifolds/Hamiltonian.jl (hunk 23 above)
Comment (kellertuer) [reply src/manifolds/Hamiltonian.jl:53]: the mathfrak one, one could add The symbol ``\mathfak{sp}`` here. (Hamiltonian itself has not that many s or p in it).

File: src/manifolds/SymplecticGrassmann.jl (hunk 25 above)
Comment (kellertuer) [reply src/manifolds/SymplecticGrassmann.jl:39]: We at least have the manifold that would be based on, HamiltonianMatrices, now, but sure, we could open an issue about that (but I will not add it here, this PR is already about a month longer than I thought).

Comment (mateuszbaran) [reply src/manifolds/SymplecticGrassmann.jl:39]: Sure, this PR is definitely not a good place for even more new stuff.

Comment (mateuszbaran) [reply src/manifolds/SymplecticGrassmann.jl:39]: I've opened an issue: #705 .

File: src/manifolds/SymplecticGrassmannStiefel.jl (hunk 28 above)
Comment (kellertuer) [reply src/manifolds/SymplecticGrassmannStiefel.jl:89]: I can check that carefully, sure.

File: src/manifolds/SymplecticGrassmann.jl (hunk 25 above)
Comment (kellertuer) [reply src/manifolds/SymplecticGrassmann.jl:39]: Oh, no! I had hoped at some point to reduce our open issues, my feeling is every PR to close one opens 2 new ones ;)

**Comment by kellertuer** [comment]
Raised tolerances before I went home, tests ran fine on my 1.10, they do not on CI – I love tolerances-guess-games – not so much.

File: src/manifolds/SymplecticGrassmann.jl (hunk 25 above)
Comment (mateuszbaran) [reply src/manifolds/SymplecticGrassmann.jl:39]: Unfortunately, it's much easier to open new issues than to close existing ones, so here we are :sweat_smile:

File: src/manifolds/SymplecticGrassmannStiefel.jl (hunk 28 above)
Comment (kellertuer) [reply src/manifolds/SymplecticGrassmannStiefel.jl:89]: Besides a missing bracket in the beginning (hehe) this looks fine. I am currently running the tests for short.

Comment (kellertuer) [reply src/manifolds/SymplecticGrassmannStiefel.jl:89]: But to be honest, this formula took me also quite some time to transfer / understand / rewrite from the paper.

File: src/manifolds/SymplecticGrassmannStiefel.jl (hunk 18 above)
Comment (kellertuer) [reply src/manifolds/SymplecticGrassmannStiefel.jl:69]: I will have to think how to best phrase that without writing half a paper.

File: src/manifolds/SymplecticGrassmannStiefel.jl (hunk 28 above)
Comment (mateuszbaran) [reply src/manifolds/SymplecticGrassmannStiefel.jl:96]: OK, let's keep your version then.

File: src/manifolds/SymplecticGrassmannStiefel.jl (hunk 18 above)
Comment (mateuszbaran) [reply src/manifolds/SymplecticGrassmannStiefel.jl]: The new version reads well enough :+1:

**Review by mateuszbaran — APPROVED** [review]
LGTM :+1:
